src/java/test/org/apache/zookeeper/server/ZooKeeperServerMainTest.java

```diff
@@ -138,14 +145,56 @@ void delete(File f) throws IOException {
         ServerCnxnFactory getCnxnFactory() {
             return main.getCnxnFactory();
         }
+
     }
 
-    public static  class TestZKSMain extends ZooKeeperServerMain {
+    public static class TestZKSMain extends ZooKeeperServerMain {
+
+        private ServerStats serverStats;
+
+        @Override
+        public ZooKeeperServer getZooKeeperServer(FileTxnSnapLog txnLog, ServerConfig config, ZKDatabase zkDb) {
+            ZooKeeperServer zooKeeperServer = super.getZooKeeperServer(txnLog, config, zkDb);
+            serverStats = zooKeeperServer.serverStats();
+            return zooKeeperServer;
+        }
+
+        @Override
         public void shutdown() {
             super.shutdown();
         }
     }
 
+    // Test for ZOOKEEPER-2770 ZooKeeper slow operation log
+    @Test
+    public void testRequestWarningThreshold() throws IOException, KeeperException, InterruptedException {
+        ClientBase.setupTestEnv();
+
+        final int CLIENT_PORT = PortAssignment.unique();
+
+        MainThread main = new MainThread(CLIENT_PORT, true, null, 0);
+        main.start();
+
+        Assert.assertTrue("waiting for server being up",
+                ClientBase.waitForServerUp("127.0.0.1:" + CLIENT_PORT,
+                        CONNECTION_TIMEOUT));
+        // Get the stats object from the ZooKeeperServer to keep track of high latency requests.
+        ServerStats stats = main.main.serverStats;
+
+        ZooKeeper zk = new ZooKeeper("127.0.0.1:" + CLIENT_PORT,
+                ClientBase.CONNECTION_TIMEOUT, this);
+
+        zk.create("/foo1", "foobar".getBytes(), Ids.OPEN_ACL_UNSAFE,
+                CreateMode.PERSISTENT);
+
+        Assert.assertEquals(new String(zk.getData("/foo1", null, null)), "foobar");
+        // It takes a while for the counter to get updated sometimes, this is added to reduce flakyness
+        Thread.sleep(1000);
```

Comment:
So if the test does any kind of operation that involves the interaction of the `FinalRequestProcessor`, the completion of that operation will occur after the counter increment has completed. The client side can then query the counter with safety.

You are correct that the counter isn't part of the state that ZK is protecting, but the effect will be the same since the `FRP` handles all requests in order.